Code looks good to me, but the failing op-e2e-HTTP-tests-l2oo are interesting:
```
goroutine 1732305 [running]:
github.com/ethereum-optimism/optimism/op-node/rollup/clsync.(*CLSync).onInvalidPayload(0xc0221bb9f0, {0x0?})
        /root/project/op-node/rollup/clsync/clsync.go:84 +0x1f
github.com/ethereum-optimism/optimism/op-node/rollup/clsync.(*CLSync).OnEvent(0xc0221bba90?, {0x3128be0?, 0x0?})
        /root/project/op-node/rollup/clsync/clsync.go:73 +0x105
github.com/ethereum-optimism/optimism/op-node/rollup.(*SynchronousDerivers).OnEvent(0xc028d9a000?, {0x3128be0, 0x0})
        /root/project/op-node/rollup/events.go:57 +0x3f
github.com/ethereum-optimism/optimism/op-node/rollup.(*SynchronousEvents).Drain(0xc0221bb950)
        /root/project/op-node/rollup/synchronous.go:72 +0x27
github.com/ethereum-optimism/optimism/op-node/rollup/driver.(*Driver).eventLoop(0xc0221bf400)
        /root/project/op-node/rollup/driver/state.go:225 +0x2fa
created by github.com/ethereum-optimism/optimism/op-node/rollup/driver.(*Driver).Start in goroutine 186395
        /root/project/op-node/rollup/driver/state.go:129 +0x185
panic: failed to read pre-image hint ack: EOF
```

I don't think I've seen tests fail like that before.

--- a/op-node/rollup/engine/engine_controller.go
+++ b/op-node/rollup/engine/engine_controller.go
@@ -266,9 +266,6 @@ func (e *EngineController) ConfirmPayload(ctx context.Context, agossip async.Asy
 	updateSafe := e.buildingSafe && e.safeAttrs != nil && e.safeAttrs.IsLastInSpan
 	envelope, errTyp, err := confirmPayload(ctx, e.log, e.engine, fc, e.buildingInfo, updateSafe, agossip, sequencerConductor)
 	if err != nil {
-		if !errors.Is(err, derive.ErrTemporary) {
-			e.emitter.Emit(InvalidPayloadEvent{})
-		}
 		return nil, errTyp, fmt.Errorf("failed to complete building on top of L2 chain %s, id: %s, error (%d): %w", e.buildingOnto, e.buildingInfo.ID, errTyp, err)
 	}
 	ref, err := derive.PayloadToBlockRef(e.rollupCfg, envelope.ExecutionPayload)

--- a/op-node/rollup/engine/events.go
+++ b/op-node/rollup/engine/events.go
@@ -24,6 +24,14 @@ func (ev InvalidPayloadEvent) String() string {
 	return "invalid-payload"
 }
 
+type InvalidPayloadAttributesEvent struct {
+	Attributes *derive.AttributesWithParent
+}
+
+func (ev InvalidPayloadAttributesEvent) String() string {
+	return "invalid-payload-attributes"
+}
+
 // ForkchoiceRequestEvent signals to the engine that it should emit an artificial
 // forkchoice-update event, to signal the latest forkchoice to other derivers.
 // This helps decouple derivers from the actual engine state,
@@ -251,6 +259,9 @@ func (eq *EngDeriver) onForceNextSafeAttributes(attributes *derive.AttributesWit
 			eq.emitter.Emit(rollup.ResetEvent{Err: fmt.Errorf("need reset to resolve pre-state problem: %w", err)})
 			return
 		case BlockInsertPayloadErr:
+			if !errors.Is(err, derive.ErrTemporary) {
+				eq.emitter.Emit(InvalidPayloadAttributesEvent{Attributes: attributes})
+			}
 			_ = eq.ec.CancelPayload(ctx, true)
 			eq.log.Warn("could not process payload derived from L1 data, dropping attributes", "err", err)
 			// Count the number of deposits to see if the tx list is deposit only.

--- a/op-program/client/driver/program.go
+++ b/op-program/client/driver/program.go
@@ -48,7 +48,7 @@ func (d *ProgramDeriver) OnEvent(ev rollup.Event) {
 		// No need to queue the attributes, since there is no unsafe chain to consolidate against,
 		// and no temporary-error retry to perform on block processing.
 		d.Emitter.Emit(engine.ProcessAttributesEvent{Attributes: x.Attributes})
-	case engine.InvalidPayloadEvent:
+	case engine.InvalidPayloadAttributesEvent:
 		// If a set of attributes was invalid, then we drop the attributes,
 		// and continue with the next.
 		d.Emitter.Emit(engine.PendingSafeRequestEvent{})

--- a/op-program/client/driver/program_test.go
+++ b/op-program/client/driver/program_test.go
@@ -73,7 +73,7 @@ func TestProgramDeriver(t *testing.T) {
 	t.Run("invalid payload", func(t *testing.T) {
 		p, m := newProgram(t, 1000)
 		m.ExpectOnce(engine.PendingSafeRequestEvent{})
-		p.OnEvent(engine.InvalidPayloadEvent{Envelope: &eth.ExecutionPayloadEnvelope{}})
+		p.OnEvent(engine.InvalidPayloadAttributesEvent{Attributes: &derive.AttributesWithParent{}})
 		m.AssertExpectations(t)
 		require.False(t, p.closing)
 		require.NoError(t, p.result)

